parley/src/layout/cursor.rs

```diff
@@ -97,304 +73,280 @@ impl Cursor {
+                if cluster.is_rtl() {
+                    clusters = [cluster.previous_visual(), Some(cluster)];
+                } else {
+                    clusters = [Some(cluster.clone()), cluster.next_visual()];
                 }
-            }
-            // Check for directional boundary condition
-            if let Some(prev) = cluster.previous_visual() {
-                if prev.is_rtl() != self.is_rtl {
-                    if let Some(prefer_rtl) = prefer_rtl {
-                        if self.is_rtl != prefer_rtl {
-                            return Self::from_cluster(prev, self.affinity.invert());
-                        }
-                    }
+            } else if let Some(cluster) = self.downstream_cluster(layout) {
+                if cluster.is_rtl() {
+                    clusters = [None, Some(cluster)];
+                } else {
+                    clusters = [Some(cluster), None];
                 }
             }
-            // We're moving left so we want to track left-side affinity;
-            // let's swap
-            Self::from_index(layout, self.index as usize, self.affinity.invert())
-        } else if let Some(prev) = cluster.previous_visual() {
-            // Handle soft line breaks
-            if matches!(
-                prev.is_line_break(),
-                Some(BreakReason::Regular) | Some(BreakReason::Emergency)
-            ) {
-                // Match the behavior of next_visual: move to the end of the soft line
-                // break
-                return Self::from_cluster(prev, self.affinity.invert());
-            }
-            let prev_rtl = prev.is_rtl();
-            // Check for directional boundary condition
-            if let Some(prev_prev) = prev.previous_visual() {
-                if prev_prev.is_rtl() != prev_rtl {
-                    if let Some(prefer_rtl) = prefer_rtl {
-                        if prev_rtl != prefer_rtl {
-                            return Self::from_cluster(prev_prev, self.affinity);
-                        } else {
-                            return Self::from_cluster(prev, self.affinity);
-                        }
-                    }
-                }
+        } else if let Some(cluster) = self.downstream_cluster(layout) {
+            if cluster.is_rtl() {
+                clusters = [Some(cluster.clone()), cluster.next_visual()];
+            } else {
+                clusters = [cluster.previous_visual(), Some(cluster)];
             }
-            let affinity = if self.is_rtl != prev_rtl {
-                self.affinity.invert()
+        } else if let Some(cluster) = self.upstream_cluster(layout) {
+            if cluster.is_rtl() {
+                clusters = [None, Some(cluster)];
             } else {
-                self.affinity
-            };
-            Self::from_cluster(prev, affinity)
-        } else {
-            *self
+                clusters = [Some(cluster), None];
+            }
```

Comment:
Might be better to hoist some of these `clusters = ` assignments.

Reply:
The repetition is a bit annoying. Fixed.